documentation/dataplane/dash-routing-actions.md

```diff
@@ -0,0 +1,197 @@
+
+## Stage transition actions
+
+### Basic stage transition actions
+
+#### `drop` action
+
+- Actions:
+  - Drop the packet after the entry is matched.
+
+#### `trap` action
+
+- Actions:
+  - Trap the packet to CPU after the entry is matched.
+
+### Matching stage transition actions
+
+All matching stage transition actions will support the following parameters:
+
+- `default_routing_type`: If no entry is found, use this routing type to route the packet. If `default_routing_type` is not set, the packet will be dropped by default.
+- `stage_index`: If there are multiple stage of the same stage type, use this field to specify which stage to transit to. The index is starting from 0 and by default 0.
+
+#### IP-based stage transition actions
+
+All IP-based stage transition actions will support the following parameters:
+
+- `use_src_ip`: Use source IP in routing and mapping stages.
+- `ip_mask`: IP mask to apply before matching the entries.
+
+##### `lpmrouting` action
+
+- Actions:
+  - The packet will be routed to the next stage based on the LPM routing table.
+
+The port mapping entries can be described as below:
+
+```json
+// Key Format: DASH_SAI_ROUTE_TABLE|<ENI ID>|<Stage Index>|<IP Cidr>
+"DASH_SAI_ROUTE_TABLE|123456789012|0|10.0.1.0/24": {
+    "transition": "some_transition",
+    "routing_type": "do_somethng",
+
+    // Metadata properties ...
+}
+```
+
+##### `maprouting` action
+
+- Actions:
+  - The packet will be routed to the next stage based on the map routing table.
+
+A map routing entry can be described as below:
+
+```json
+// Key Format: DASH_VNET_MAPPING_TABLE|<VNET ID>|<Stage Index>|<IP>
+"DASH_VNET_MAPPING_TABLE|Vnet1|0|10.0.1.1": {
+    "transition": "some_transition",
+    "routing_type": "do_somethng",
+
+    // Metadata properties ...
+}
+```
+
+#### Port-based stage transition actions
+
+##### `portmaprouting` action
+
+- Actions:
+  - The packet will be routed to the next stage with TCP or UDP port mapping.
+
+The port mapping entries can be described as below:
+
+```json
+// Key Format: DASH_SAI_PORT_MAPPING_TABLE|<Port Mapping Id>
+"DASH_SAI_PORT_MAPPING_TABLE|portmapping-0": [
+    // ...
+    {
+        "transition": "some_transition",
+        "routing_type": "do_something",
+
+        // Port range to match
+        "src_port_min": 0,
+        "src_port_max": 65535,
+        "dst_port_min": 2000,
+        "dst_port_max": 2099,
+
+        // Metadata properties ...
+    }
+    // ...
+]
+```
+
+## Packet transformation actions
+
+### `staticencap` action
+
+- Metadata parameters:
+  - `underlay_dip`: Destination IP used in encap.
+  - `underlay_sip`: Source IP used in encap.
+  - `encap_type`: "nvgre|vxlan|…"
+  - `encap_key`: GRE key in NvGRE or VNI in VxLAN
+- Actions:
+  - Enable the underlay encap header based on the encap type.
+  - Update the underlay encap header with `encap_key`, `underlay_dip`, `underlay_sip`.
+    - If `underlay_dip` / `underlay_sip` is not set, use the original IP.
```

Comment:
If we are encapsulating a tenant packet and let's say only underlay_dip is set in the action, then are we supposed to copy tenant(overlay) sip to underlay_sip? Then packet will go out with an encap consisting of dip from the underlay address range and sip from the overlay address range. Doesn't sound right to me.

Reply:
The pipeline has the `underlay_ip` field, which will be published as `underlay_sip` in outbound pipeline and `underlay_dip` as inbound pipeline. 

I forgot to removed the last item and it is not removed to avoid confusion.